Fix Bulk Insert: Single Column and MySQL earlier upsert syntax

QuangTung97 opened the pull request:
This change handle the case of using single column in "VALUES(...)".
And allow using VALUES in "INSERT ... ON DUPLICATE KEY UPDATE ..." kind of queries,  e.g issue https://github.com/jmoiron/sqlx/issues/722
Possible fixes for https://github.com/jmoiron/sqlx/issues/701 https://github.com/jmoiron/sqlx/issues/695 https://github.com/jmoiron/sqlx/issues/694

jakeaglass commented:
This solves issues I was having with the new batch insert feature, which was generating statements with unbalanced parentheses, eg `INSERT (id) VALUES (coalesce(:a,:b))` compiled to `INSERT (id) VALUES (coalesce(1,2)`

Many thanks @QuangTung97

w1ck3dg0ph3r commented:
Thanks @QuangTung97!
Was about to create similar PR that does away with this brittle regex)

By the way, this should also fix some additional cases:
- Functions inside values: `INSERT INTO foo (a,b,c,d) VALUES (:name, cast(:age as int), :first, :last)`
- Some strange regex indentation failure I was having on multi-line queries (regex failed on some levels of indentation and not the others, go figure).

Could you add those test cases as well, please? Here's the gist of the ones I added: https://gist.github.com/w1ck3dg0ph3r/cfbce1590fdce1ce54a179b6cb74f6a4.

QuangTung97 commented:
> Thanks @QuangTung97!
> Was about to create similar PR that does away with this brittle regex)
> 
> By the way, this should also fix some additional cases:
> 
> * Functions inside values: `INSERT INTO foo (a,b,c,d) VALUES (:name, cast(:age as int), :first, :last)`
> * Some strange regex indentation failure I was having on multi-line queries (regex failed on some levels of indentation and not the others, go figure).
> 
> Could you add those test cases as well, please? Here's the gist of the ones I added: https://gist.github.com/w1ck3dg0ph3r/cfbce1590fdce1ce54a179b6cb74f6a4.

Thanks. I added test for the multiple lines query case. But the case of function calls might be redundant with the test case `two level depth function call` I already added.

jmoiron commented:
This looks great;  this regex has seen a lot of authors and is difficult to get right, but this looks great and more tests are also welcome.  Going to cut a new dot release w/ this sometime this weekend, triaging more in the backlog.

w1ck3dg0ph3r commented on named.go:
```diff
@@ -224,29 +224,47 @@ func bindStruct(bindType int, query string, arg interface{}, m *reflectx.Mapper)
 }
 
-var valueBracketReg = regexp.MustCompile(`(?i)VALUES\s*(\([^(]*.[^(]\))`)
+var valuesReg = regexp.MustCompile(`\)\s*(?i)VALUES\s*\(`)
```
Should we also care for inserts without an explicit column list here, e.g. `insert into tblname values (...)`?

QuangTung97 replied:
Oh. Thanks for pointing out. But I dont find a good way to both satisfying that case and the case when the name of table containing suffix: "values".

w1ck3dg0ph3r replied:
I think `\W+(?i)VALUES\s*\(` should handle most cases, including:
```
INSERT INTO foo_values (a,b,c,d) VALUES (?, ?, ?, ?)
INSERT INTO foo_values VALUES (?, ?, ?, ?)
INSERT INTO `values`VALUES (?, ?, ?, ?)
```
While generally being a bad idea, I have no doubt that somebody does implement it, given such a wide adoption of sqlx)
